### PR TITLE
Create a plugin for IPython and improve utils.run

### LINKED ISSUE #19: [FEATURE] Create comeback plugin for IPython
Nomnomnom

### PR DESCRIPTION
**Pull request by yammesicka**
Closes #19

### CODE REVIEW

File: comeback/plugins/vscode/main.py [hunk 1]
```diff
@@ -10,7 +10,7 @@ def check_plugin(cwd: Optional[str] = None) -> utils.RUN_STATUS:
     if 'cwd' is None:
         return False, 'cwd parameter is not set.'
 
-    if not utils.binary_exists('code'):
+    if not utils.is_binary_exists('code'):
```
Comment (agamm) [inline comeback/plugins/vscode/main.py:13]: You forgot to change it in other places (pycharm plugin)

File: comeback/utils.py [hunk 2]
```diff
@@ -19,17 +20,36 @@ def get_platform() -> str:
...
 
 
-def run(cmd: str, wait: bool = False,
-         use_shell: bool = False) -> None:
-    # Create a detached windows process
-    if wait:
-        subprocess.run(cmd)
-        return
+def _format_command(cmd: Union[str, List[str]]) -> List[str]:
+    if isinstance(cmd, str):
```
Comment (agamm) [inline comeback/utils.py:24]: what if it's not a string?

File: comeback/utils.py [hunk 3]
```diff
@@ -19,17 +20,36 @@ def get_platform() -> str:
...
+        cmd = shlex.split(cmd)
+    return cmd
 
-    subprocess.Popen(cmd, shell=use_shell)
 
+def _get_detach_flags() -> Dict[str, Any]:
+    kwargs = {}
+    if get_platform() == 'windows':
+        CREATE_NEW_PROCESS_GROUP = 0x00000200
```
Comment (agamm) [inline comeback/utils.py:33]: Ohhh no! windows 😨

File: comeback/utils.py [hunk 4]
```diff
@@ -19,17 +20,36 @@ def get_platform() -> str:
...
+        kwargs['creationflags'] = DETACHED_PROCESS | CREATE_NEW_PROCESS_GROUP
+    else:
+        kwargs['start_new_session'] = True
+    return kwargs
 
-def binary_exists(bin_name: str) -> bool:
+
+def run(cmd: Union[List[str], str],
+        use_shell: bool = False,
+        detach: bool = True) -> None:
```
Comment (agamm) [inline comeback/utils.py]: Add a new line maybe? it looks like it's part of the code

File: comeback/utils.py [hunk 5]
```diff
@@ -19,17 +20,36 @@ def get_platform() -> str:
...
+    cmd = _format_command(cmd)
+
+    kwargs = {}
```
Comment (agamm) [inline comeback/utils.py:44]: Maybe a less generic name? it's not specifically for the detach on windows?

File: example/example_yamm/.comeback [hunk 6]
```diff
@@ -0,0 +1,6 @@
+ipython:
+  modules:
+    - pathlib
+    - random
+    - this
+    - typing
```
Comment (agamm) [inline example/example_yamm/.comeback:6]: what about `import antigravity`?

File: comeback/utils.py [hunk 7]
```diff
@@ -19,17 +20,36 @@ def get_platform() -> str:
     return platforms.get(platform.system(), 'other')
 
 
-def run(cmd: str, wait: bool = False,
-         use_shell: bool = False) -> None:
-    # Create a detached windows process
-    if wait:
-        subprocess.run(cmd)
-        return
+def _format_command(cmd: Union[str, List[str]]) -> List[str]:
```
Comment (agamm) [inline comeback/utils.py]: union is coooowel.

File: comeback/plugins/vscode/main.py (hunk 1 above)
Comment (yammesicka) [reply comeback/plugins/vscode/main.py:13]: Fixed it in the `PyCharm` commit

File: comeback/utils.py (hunk 7 above)
Comment (yammesicka) [reply comeback/utils.py]: Yeah 😎

File: comeback/utils.py (hunk 3 above)
Comment (yammesicka) [reply comeback/utils.py:33]: LOL XDXD

File: comeback/utils.py (hunk 5 above)
Comment (yammesicka) [reply comeback/utils.py:44]: Not specifically for win - it's for detach. There is a special case for Linux too.
\+ In the future we might want to add some kwargs

File: example/example_yamm/.comeback (hunk 6 above)
Comment (yammesicka) [reply example/example_yamm/.comeback:6]: Try `from __future__ import barry_as_flufl` :PPP

File: comeback/utils.py (hunk 2 above)
Comment (yammesicka) [reply comeback/utils.py:24]: So I assume it's probably already a `List` :)

**Review by agamm — APPROVED** [review]

**Pull request merged**